File: python/dumpscript.py
import sys
import csv
import traceback
import time
import psycopg2
from pymongo import MongoClient
from collections import defaultdict
import argparse

def print_usage():
    """Print usage instructions."""
    print("Usage: python3 dumpscript.py -f filename.ext -t [mongo|postgres]")
    print("Options:")
    print("  -f, --file     Specify the CSV file to read")
    print("  -t, --type     Specify the type of database to use (mongo or postgres)")
    print("  --help         Display this help message")

def process_csv(filename):
    """Process CSV file and determine column data types."""
    column_data_types = defaultdict(set)
    try:
        with open(filename, 'r') as file:
            reader = csv.reader(file)
            columns = next(reader)
            for row in reader:
                for i, value in enumerate(row):
                    try:
                        int(value)
                        column_data_types[columns[i]].add('int')
                    except ValueError:
                        try:
                            float(value)
                            if 'int' not in column_data_types[columns[i]]:
                                column_data_types[columns[i]].add('float')
                        except ValueError:
                            column_data_types[columns[i]].add('varchar')
    except Exception as e:
        print(f"Error processing CSV: {e}")
        traceback.print_exc()
    return columns, column_data_types

def create_postgres_table(cur, table_name, column_data_types):
    """Create PostgreSQL table."""
    create_table_query = f"CREATE TABLE IF NOT EXISTS {table_name} ("
    for column, data_types in column_data_types.items():
        if 'varchar' in data_types:
            data_types.discard('float')
            data_types.discard('int')
        elif 'float' in data_types:
            data_types.discard('int')
        elif len(data_types) > 1 and 'int' in data_types:
            data_types.discard('int')
        data_type = ' '.join(data_types)
        create_table_query += f"{column} {data_type}, "
    create_table_query = create_table_query[:-2] + ")"
    cur.execute(create_table_query)

def insert_into_postgres(filename, table_name, columns, column_data_types):
    """Insert data into PostgreSQL."""
    try:
        conn = psycopg2.connect(
            dbname="postgres",
            user="postgres",
            password="lfkoP@ssw0rd",
            host="localhost",
            port="5432"
        )
        cur = conn.cursor()

        create_postgres_table(cur, table_name, column_data_types)
        conn.commit()

        start_time = time.time()
        total_rows = 0
        rows_added = 0
        errors = 0
        with open(filename, 'r') as file:
            reader = csv.reader(file)
            next(reader)
            for row in reader:
                total_rows += 1
                try:
                    row = [int(float(cell)) if '.' in cell and cell.replace('.', '', 1).isdigit() else float(cell) if cell.replace('.', '', 1).isdigit() else cell for cell in row]
                    cur.execute(
                        f"INSERT INTO {table_name} VALUES ({','.join(['%s'] * len(row))})",
                        row
                    )
                    rows_added += 1
                except ValueError as e:
                    errors += 1
                    print(f"Error inserting row {total_rows}: {e}")
                    continue

        conn.commit()
        cur.close()
        conn.close()

        elapsed_time = time.time() - start_time
        print("Table created successfully in PostgreSQL!")
        print("Data inserted successfully into PostgreSQL!")
        print(f"Time elapsed: {elapsed_time:.2f} seconds")
        print(f"Total rows processed: {total_rows}")
        print(f"Total rows added: {rows_added}")
        print(f"Total errors encountered: {errors}")
    except Exception as e:
        print(f"Error inserting into PostgreSQL: {e}")
        traceback.print_exc()

def insert_into_mongo(filename, table_name):
    """Insert data into MongoDB."""
    try:
        client = MongoClient('mongodb://localhost:27017/')
        db = client['your_database_name']
        collection = db[table_name]

        start_time = time.time()
        total_rows = 0
        rows_added = 0
        errors = 0
        inserted_ids = []
        with open(filename, 'r') as file:
            reader = csv.DictReader(file)
            for row in reader:
                total_rows += 1
                try:
                    result = collection.insert_one(row)
                    inserted_ids.append(result.inserted_id)
                    rows_added += 1
                except Exception as e:
                    errors += 1
                    print(f"Error inserting row {total_rows}: {e}")

        elapsed_time = time.time() - start_time
        print("Data inserted successfully into MongoDB!")
        print(f"Time elapsed: {elapsed_time:.2f} seconds")
        print(f"Total rows processed: {total_rows}")
        print(f"Total rows added: {rows_added}")
        print(f"Total errors encountered: {errors}")
    except Exception as e:
        print(f"Error inserting into MongoDB: {e}")
        traceback.print_exc()

def main():
    """Main function."""
    parser = argparse.ArgumentParser(description="Dump CSV data into a database.")
    parser.add_argument('-f', '--file', type=str, help="CSV file to read", required=True)
    parser.add_argument('-t', '--type', type=str, help="Type of database to use (mongo or postgres)", required=True)
    args = parser.parse_args()

    filename = args.file
    db_type = args.type

    columns, column_data_types = process_csv(filename)
    table_name = filename.split('/')[-1].split('.')[0]

    if db_type == 'postgres':
        insert_into_postgres(filename, table_name, columns, column_data_types)
    elif db_type == 'mongo':
        insert_into_mongo(filename, table_name)
    else:
        print("Invalid database type. Supported types are 'mongo' and 'postgres'.")

if __name__ == "__main__":
    main()

File: node/dumpscript.js
const fs = require("fs");
const csv = require("csv-parser");
const { Pool } = require("pg");
const { MongoClient } = require("mongodb");

// PostgreSQL connection pool
const pool = new Pool({
  user: "postgres",
  host: "localhost",
  database: "postgres",
  password: "lfkoP@ssw0rd",
  port: 5432,
  max: 20, // Increase the maximum number of clients in the pool
});

// MongoDB connection URL
const mongoURL = "mongodb://localhost:27017";
const dbName = "your_database_name";

const processCSV = async (filename) => {
  const columnDataTypes = [];
  let columns = [];
  try {
    const stream = fs.createReadStream(filename).pipe(csv());
    stream.on("headers", (headers) => {
      columns = headers;
    });
    stream.on("data", (data) => {
      columns.forEach((column, index) => {
        const value = data[column];
        let dataType;
        if (!isNaN(value)) {
          if (Number.isInteger(Number(value))) {
            dataType = "int";
          } else {
            dataType = "float";
          }
        } else {
          dataType = "varchar";
        }
        if (!columnDataTypes[index]) {
          columnDataTypes[index] = dataType;
        } else if (columnDataTypes[index] !== dataType) {
          columnDataTypes[index] = "varchar";
        }
      });
    });
    await new Promise((resolve, reject) => {
      stream.on("end", resolve);
      stream.on("error", reject);
    });
  } catch (error) {
    console.error(`Error processing CSV: ${error}`);
    throw error;
  }
  console.log("Processed CSV successfully", { columns, columnDataTypes });
  return { columns, columnDataTypes };
};

const insertIntoPostgres = async (
  filename,
  tableName,
  columns,
  columnDataTypes
) => {
  const client = await pool.connect();
  try {
    let createTableQuery = `CREATE TABLE IF NOT EXISTS ${tableName} (`;
    columns.forEach((column, index) => {
      createTableQuery += `${column} ${columnDataTypes[index]}, `;
    });
    createTableQuery = createTableQuery.slice(0, -2) + ")";
    await client.query(createTableQuery);

    const start = Date.now();
    let totalRows = 0;
    let rowsAdded = 0;
    let errors = 0;

    const stream = fs.createReadStream(filename).pipe(csv());
    for await (const data of stream) {
      totalRows++;
      const values = columns.map((column) => data[column]);
      console.log("Inserting row", { values });

      try {
        await client.query(
          `INSERT INTO ${tableName} (${columns.join(", ")}) VALUES (${values
            .map((_, index) => `$${index + 1}`)
            .join(", ")})`,
          values
        );
        rowsAdded++;
      } catch (error) {
        errors++;
        console.error(`Error inserting row ${totalRows}: ${error}`);
      }
    }

    const elapsedTime = (Date.now() - start) / 1000;
    console.log("Data inserted successfully into PostgreSQL", {
      elapsedTime,
      totalRows,
      rowsAdded,
      errors,
    });
  } catch (error) {
    console.error(`Error inserting into PostgreSQL: ${error}`);
    throw error;
  } finally {
    client.release();
  }
};

const insertIntoMongoDB = async (filename, collectionName) => {
  const client = new MongoClient(mongoURL);

  try {
    await client.connect();
    console.log("Connected to MongoDB successfully!");

    const db = client.db(dbName);
    const collection = db.collection(collectionName);

    const start = Date.now();
    let totalRows = 0;
    let documentsInserted = 0;
    let errors = 0;

    const stream = fs.createReadStream(filename).pipe(csv());
    for await (const data of stream) {
      totalRows++;
      console.log("Inserting document", { data });

      try {
        await collection.insertOne(data);
        documentsInserted++;
      } catch (error) {
        errors++;
        console.error(`Error inserting document ${totalRows}: ${error}`);
      }
    }

    const elapsedTime = (Date.now() - start) / 1000;
    console.log("Data inserted successfully into MongoDB", {
      elapsedTime,
      totalRows,
      documentsInserted,
      errors,
    });
  } catch (error) {
    console.error(`Error inserting into MongoDB: ${error}`);
    throw error;
  } finally {
    await client.close();
  }
};

const main = async () => {
  try {
    // Process command line arguments
    const args = process.argv.slice(2);
    // Your logic for handling arguments goes here

    // Example usage:
    const filename =
      args[args.indexOf("-f") + 1] || args[args.indexOf("--file") + 1];
    const tableName =
      args[args.indexOf("-t") + 1] || args[args.indexOf("--table") + 1];
    const dbType =
      args[args.indexOf("-t") + 1] || args[args.indexOf("--type") + 1];

    // Process CSV and insert into PostgreSQL
    const { columns, columnDataTypes } = await processCSV(filename);

    if (dbType === "postgres") {
      await insertIntoPostgres(filename, tableName, columns, columnDataTypes);
    } else if (dbType === "mongo") {
      await insertIntoMongoDB(filename, tableName);
    } else {
      console.log(
        "Invalid database type. Supported types are 'mongo' and 'postgres'."
      );
    }

    console.log("Script execution completed successfully!");
  } catch (error) {
    console.error(`Script execution failed: ${error}`);
    process.exit(1); // Exit with non-zero status code
  }
};

main();
